engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroToJsonFormatterFactory.scala

```diff
@@ -0,0 +1,119 @@
+    * @tparam V - value type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+    */
+  protected def doFormatRecord[K: ClassTag, V: ClassTag](kafkaConfig: KafkaConfig,
+                                                         kafkaSourceDeserializationSchema: KafkaDeserializationSchema[ConsumerRecord[K, V]],
+                                                         messageFormatter: ConfluentAvroMessageFormatter)
+                                                        (record: ConsumerRecord[Array[Byte], Array[Byte]]): Array[Byte] = {
+    val deserializedRecord = kafkaSourceDeserializationSchema.deserialize(record)
+    val serializableRecord = AvroSerializableConsumerRecord(
+      messageFormatter.getSchemaIdOpt(record.key()),
+      messageFormatter.getSchemaIdOpt(record.value()).get,
+      SerializableConsumerRecord(deserializedRecord)
+    )
+
+    implicit val keyEncoder: Encoder[K] = createKeyEncoder(serializableRecord.consumerRecord.key.get, kafkaConfig.useStringForKey, messageFormatter) // TODO: key.get???
+    implicit val valueEncoder: Encoder[V] = createValueEncoder(serializableRecord.consumerRecord.value, messageFormatter)
+    implicit val serializableRecordEncoder: Encoder[SerializableConsumerRecord[K, V]] = deriveEncoder
+    val consumerRecordEncoder: Encoder[AvroSerializableConsumerRecord[K, V]] = deriveEncoder
+
+    consumerRecordEncoder(serializableRecord).noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  /**
+    * Step 1: Deserialize raw json bytes to AvroSerializableConsumerRecord[Json, Json] domain without interpreting key and value content.
+    * Step 2: Create key and value json-to-avro interpreter based on schema id's provided in json.
+    * Step 3: Use interpreter to create raw kafka ConsumerRecord
+    * @tparam K - key type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+    * @tparam V - value type passed from KafkaAvroSourceFactory, used to determine which datumReaderWriter use (e.g. specific or generic)
+    */
+  protected def doParseRecord[K: ClassTag, V: ClassTag](kafkaConfig: KafkaConfig,
+                                                        createReader: String => ConfluentAvroMessageReader)
+                                                       (topic: String, bytes: Array[Byte]): ConsumerRecord[Array[Byte], Array[Byte]] = {
+    implicit val serializableRecordDecoder: Decoder[SerializableConsumerRecord[Json, Json]] = deriveDecoder
+    val consumerRecordDecoder: Decoder[AvroSerializableConsumerRecord[Json, Json]] = deriveDecoder
+    val record = CirceUtil.decodeJsonUnsafe(bytes)(consumerRecordDecoder)
+
+    def serializeKeyValue(keyOpt: Option[Json], value: Json): (Array[Byte], Array[Byte]) = {
+      val reader = createReader(topic)
+      val keyBytes = if (kafkaConfig.useStringForKey) {
```

Comment:
I don't exactly understand why this code has such many cases. when `useStringForKey` is true, shouldn't you expect that keyOpt is always  `Some(j) if j.isString`?

Reply:
I see it necessary to handle null and emptyByteArray key values.